### PR TITLE
Add tests for fake URL creator and retriever

### PR DESCRIPTION
**Pull request by KamalashreeNagaraj**
Issue-id: #197

Signed-off-by: Kamalashree N <nagaraj.kamalashree@gmail.com>

### Description
Improve test-coverage for `app/usecase/url` package.

### CODE REVIEW

**Comment by KamalashreeNagaraj** [comment]
@byliuyang, will you be reviewing this commit?

**Comment by byliuyang** [comment]
@KamalashreeNagaraj I am working on it.

File: app/usecase/url/urlretriever_fake_test.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	"time"
+
+	"github.com/byliuyang/app/mdtest"
+)
+
+type fakeUrlMap = map[string]entity.URL
+
+func TestFakeUrlRetriever_GetUrlAfter(t *testing.T) {
+	now := time.Now()
+	before := now.Add(-5 * time.Second)
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
	before := now.Add(time.Second)
```

File: app/usecase/url/urlcreator_fake_test.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+				Alias:    "220uFicCJj",
+				ExpireAt: &now,
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			fakeCreator := NewCreatorFake(testCase.urls, testCase.alias)
+			url, err := fakeCreator.Create(testCase.url, testCase.userEmail)
```
Comment (byliuyang) [inline app/usecase/url/urlcreator_fake_test.go]:
```suggestion
			gotURL, err := fakeCreator.Create(testCase.url, testCase.userEmail)
```

File: app/usecase/url/urlcreator_fake_test.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+			if testCase.hasErr {
+				mdtest.NotEqual(t, nil, err)
+				return
+			}
+			mdtest.Equal(t, nil, err)
+			mdtest.Equal(t, testCase.expectedURL, url)
```
Comment (byliuyang) [inline app/usecase/url/urlcreator_fake_test.go]:
```suggestion
			mdtest.Equal(t, testCase.expectedURL, gotURL)
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			tracer := mdtest.NewTracerFake()
+			fakeTrace := tracer.BeginTrace("GetUrlAfter")
+			fakeRetriever := NewRetrieverFake(testCase.urls)
+
+			url, err := fakeRetriever.GetAfter(fakeTrace, testCase.alias, testCase.expiringAt)
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
			gotURL, err := fakeRetriever.GetAfter(fakeTrace, testCase.alias, testCase.expiringAt)
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+			if testCase.hasErr {
+				mdtest.NotEqual(t, nil, err)
+				return
+			}
+			mdtest.Equal(t, nil, err)
+			mdtest.Equal(t, testCase.expectedURL, url)
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
			mdtest.Equal(t, testCase.expectedURL, gotURL)
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			tracer := mdtest.NewTracerFake()
+			fakeTrace := tracer.BeginTrace("GetUrl")
+			fakeRetriever := NewRetrieverFake(testCase.urls)
+
+			url, err := fakeRetriever.Get(fakeTrace, testCase.alias)
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
			gotURL, err := fakeRetriever.Get(fakeTrace, testCase.alias)
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+			if testCase.hasErr {
+				mdtest.NotEqual(t, nil, err)
+				return
+			}
+			mdtest.Equal(t, nil, err)
+			mdtest.Equal(t, testCase.expectedURL, url)
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
			mdtest.Equal(t, testCase.expectedURL, gotURL)
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+				Alias:    "220uFicCJj",
+				ExpireAt: &now,
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			tracer := mdtest.NewTracerFake()
+			fakeTrace := tracer.BeginTrace("GetUrl")
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
			fakeTrace := tracer.BeginTrace("GetURL")
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+				Alias:    "220uFicCJj",
+				ExpireAt: &after,
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			tracer := mdtest.NewTracerFake()
+			fakeTrace := tracer.BeginTrace("GetUrlAfter")
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
			fakeTrace := tracer.BeginTrace("GetURLAfter")
```

**Comment by byliuyang** [comment]
Hi @KamalashreeNagaraj, thank you for working on this PR. Please address the comments so that we can merge it.

**Comment by KamalashreeNagaraj** [comment]
Sure, will do 👍

File: app/usecase/url/urlretriever_fake_test.go (hunk 1 above)
Comment (KamalashreeNagaraj) [reply app/usecase/url/urlretriever_fake_test.go]: You ideally want to subtract right? Why are we removing the negation?

**Comment by byliuyang** [comment]
Sorry, I miss read the variable name. Subtraction is preferred.

**Comment by KamalashreeNagaraj** [comment]
Also, I see this error string `"usecase: URL alias already exist"`. Is the word `usecase` intended?

**Comment by KamalashreeNagaraj** [comment]
> Sorry, I miss read the variable name. Subtraction is preferred.
> […](#)
> On Oct 8, 2019, at 10:19, Kamalashree Nagaraj ***@***.***> wrote: @KamalashreeNagaraj commented on this pull request. In app/usecase/url/urlretriever_fake_test.go: > @@ -0,0 +1,149 @@ +package url + +import ( + "short/app/entity" + "testing" + "time" + + "github.com/byliuyang/app/mdtest" +) + +type fakeUrlMap = map[string]entity.URL + +func TestFakeUrlRetriever_GetUrlAfter(t *testing.T) { + now := time.Now() + before := now.Add(-5 * time.Second) You ideally want to subtract right? Why are we removing the negation? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

The `time` package does not expose a Subtract method that takes duration as a parameter. It expects a `Time` type as the parameter.
I tried extracting a private function as this 
```
func sub(from time.Time, duration time.Duration) time.Time {
	return from.Add(-duration)
}
```
Not sure how much this helps. Let me know your comments on this.

**Comment by byliuyang** [comment]
Hi @KamalashreeNagaraj , can you link the line of error message here so that I can look it up?

I agree that the ‘sub’ function is overkill in this case.

**Comment by KamalashreeNagaraj** [comment]
> Hi @KamalashreeNagaraj , can you link the line of error message here so that I can look it up?
> 
> I agree that the ‘sub’ function is overkill in this case.

https://github.com/byliuyang/short/blob/master/app/usecase/url/urlcreator_fake.go#L25

**Comment by byliuyang** [comment]
@KamalashreeNagaraj I originally used `usecase` to make the location of error. We can remove it for now and open an issue to configure stack trace!

File: app/usecase/url/urlretriever_fake_test.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+		t.Run(testCase.name, func(t *testing.T) {
+			tracer := mdtest.NewTracerFake()
+			fakeTrace := tracer.BeginTrace("GetURL")
+			fakeRetriever := NewRetrieverFake(testCase.urls)
+
+			gotURL, err := fakeRetriever.Get(fakeTrace, testCase.alias)
+
+			if testCase.hasErr {
+				mdtest.NotEqual(t, nil, err)
+				mdtest.Equal(t, testCase.expectedErr, err.Error())
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+		t.Run(testCase.name, func(t *testing.T) {
+			tracer := mdtest.NewTracerFake()
+			fakeTrace := tracer.BeginTrace("GetURLAfter")
+			fakeRetriever := NewRetrieverFake(testCase.urls)
+
+			gotURL, err := fakeRetriever.GetAfter(fakeTrace, testCase.alias, testCase.expiringAt)
+
+			if testCase.hasErr {
+				mdtest.NotEqual(t, nil, err)
+				mdtest.Equal(t, testCase.expectedErr, err.Error())
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
```

File: app/usecase/url/urlcreator_fake_test.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+	testCases := []struct {
+		name        string
+		urls        urlMap
+		alias       []string
+		userEmail   string
+		url         entity.URL
+		hasErr      bool
+		expectedURL entity.URL
+		expectedErr string
```
Comment (byliuyang) [inline app/usecase/url/urlcreator_fake_test.go]:
```suggestion
```

File: app/usecase/url/urlcreator_fake_test.go [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+				"220uFicCJj": entity.URL{
+					Alias:    "220uFicCJj",
+					ExpireAt: &now,
+				},
+			},
+			alias:       []string{"220uFicCJj"},
+			userEmail:   "alpha@example.com",
+			url:         entity.URL{},
+			hasErr:      true,
+			expectedErr: "usecase: url alias already exist",
```
Comment (byliuyang) [inline app/usecase/url/urlcreator_fake_test.go]:
```suggestion
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+	after := now.Add(5 * time.Second)
+
+	testCases := []struct {
+		name        string
+		urls        fakeURLMap
+		alias       string
+		expiringAt  time.Time
+		hasErr      bool
+		expectedURL entity.URL
+		expectedErr string
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+	}{
+		{
+			name:        "alias not found",
+			urls:        fakeURLMap{},
+			alias:       "220uFicCJj",
+			expiringAt:  now,
+			hasErr:      true,
+			expectedURL: entity.URL{},
+			expectedErr: "url not found",
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+				"220uFicCJj": entity.URL{
+					Alias:    "220uFicCJj",
+					ExpireAt: &before,
+				},
+			},
+			alias:       "220uFicCJj",
+			expiringAt:  now,
+			hasErr:      true,
+			expectedURL: entity.URL{},
+			expectedErr: "url expired",
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+func TestFakeURLRetriever_GetURL(t *testing.T) {
+	now := time.Now()
+
+	testCases := []struct {
+		name        string
+		urls        fakeURLMap
+		alias       string
+		hasErr      bool
+		expectedURL entity.URL
+		expectedErr string
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
```

File: app/usecase/url/urlretriever_fake_test.go [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+	}{
+		{
+			name:        "alias not found",
+			urls:        fakeURLMap{},
+			alias:       "220uFicCJj",
+			hasErr:      true,
+			expectedURL: entity.URL{},
+			expectedErr: "url not found",
```
Comment (byliuyang) [inline app/usecase/url/urlretriever_fake_test.go]:
```suggestion
```

**Comment by byliuyang** [comment]
The master of Go said: "As an aside, I believe you should never inspect the output of the error.Error method. The Error method on the error interface exists for humans, not code."

https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

**Comment by KamalashreeNagaraj** [comment]
> The master of Go said: "As an aside, I believe you should never inspect the output of the error. Error method. The Error method on the error interface exists for humans, not code."
> 
> https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

Oh. Wasn't aware. That was a learning for me. Thank you. Will fix this.

**Comment by byliuyang** [comment]
> The master of Go said: "As an aside, I believe you should never inspect the output of the error. Error method. The Error method on the error interface exists for humans, not code."
> 
> https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

**Pull request merged**